Userland/Libraries/LibWasm/Parser/Parser.cpp

```diff
@@ -223,50 +230,48 @@ ParseResult<TableType> TableType::parse(InputStream& stream)
     return TableType { type_result.release_value(), limits_result.release_value() };
 }
 
-ParseResult<GlobalType> GlobalType::parse(InputStream& stream)
+ParseResult<GlobalType> GlobalType::parse(Core::Stream::Stream& stream)
 {
     ScopeLogger<WASM_BINPARSER_DEBUG> logger("GlobalType"sv);
     auto type_result = ValueType::parse(stream);
     if (type_result.is_error())
         return type_result.error();
-    u8 mutable_;
-    stream >> mutable_;
 
-    if (stream.has_any_error())
+    auto mutable_or_error = stream.read_value<u8>();
+    if (mutable_or_error.is_error())
         return with_eof_check(stream, ParseError::ExpectedKindTag);
 
+    auto mutable_ = mutable_or_error.release_value();
+
     if (mutable_ > 1)
         return with_eof_check(stream, ParseError::InvalidTag);
 
     return GlobalType { type_result.release_value(), mutable_ == 0x01 };
 }
 
-ParseResult<BlockType> BlockType::parse(InputStream& stream)
+ParseResult<BlockType> BlockType::parse(Core::Stream::Stream& stream)
 {
     ScopeLogger<WASM_BINPARSER_DEBUG> logger("BlockType"sv);
-    u8 kind;
-    stream >> kind;
-    if (stream.has_any_error())
+    auto kind_or_error = stream.read_value<u8>();
+    if (kind_or_error.is_error())
         return with_eof_check(stream, ParseError::ExpectedKindTag);
+
+    auto kind = kind_or_error.release_value();
     if (kind == Constants::empty_block_tag)
         return BlockType {};
 
     {
-        InputMemoryStream value_stream { ReadonlyBytes { &kind, 1 } };
-        if (auto value_type = ValueType::parse(value_stream); !value_type.is_error())
+        auto value_stream = Core::Stream::FixedMemoryStream::construct(ReadonlyBytes { &kind, 1 }).release_value_but_fixme_should_propagate_errors();
```

Comment:
hmm...I don't love the fact that we're allocating streams like crazy now, but it's _probably_ not a concern...
could you do a little perf test? just the runtime of `test-wasm` with the spec tests should be good enough I think.

Reply:
It doesn't look like it makes much of a difference, both the old and the new implementation take somewhere between 33.5 and 34.5 seconds on a warm run, which should be well within margin of error since I'm not running the tests bare-metal.

That said, looking into removing the fallible `FixedMemoryStream` constructor (there is nothing fallible in there) has been somewhere on my list anyways, so we might get some performance improvements down the line if we allow users to not allocate the stream on the heap.